wp/lib/bap_wp/src/environment.ml

```diff
@@ -125,6 +130,41 @@ let wp_rec_call :
   (t -> Constr.t -> start:Graphs.Ir.Node.t -> Graphs.Ir.t -> t) ref =
   ref (fun _ _ ~start:_ _ -> assert false)
 
+let trivial_pre (env : t) : Constr.t =
+  let ctx = get_context env in
+  Z3.Boolean.mk_true ctx
+  |> Constr.mk_goal "true"
+  |> Constr.mk_constr
+
+let post_dominator (env : t) (node : Graphs.Ir.Node.t) (graph : Graphs.Ir.t)
+  : Constr.t option =
+  let module G = Graphlib.Std in
+  let module Node = Graphs.Ir.Node in
+  let scc = G.Graphlib.strong_components (module Graphs.Ir) graph in
+  match G.Partition.group scc node with
+  | None -> None
+  | Some g ->
+    let nodes = G.Group.enum g in
+    let leaf = Seq.find (G.Graphlib.postorder_traverse (module Graphs.Ir) graph)
+        ~f:(fun n -> Seq.is_empty (Node.succs n graph))
+    in
+    match leaf with
+    | None -> None
+    | Some l ->
+      let dom_tree = G.Graphlib.dominators (module Graphs.Ir) ~rev:true graph l in
+      let exit_node = Seq.find_map nodes ~f:(fun n ->
+          match G.Tree.parent dom_tree n with
+          | None -> None
+          | Some p ->
+            if Seq.exists nodes ~f:(Node.equal p) then
+              None
+            else
+              Some (p |> Node.label |> Term.tid)
```

Comment:
This feels a bit backwards, can't we directly ask whether `p` is not in `g`, say by using `Group.mem`?